### PR TITLE
Upgrade to winstone with Jetty 12

### PR DESCRIPTION
**Pull request by olamy**
Signed-off-by: Olivier Lamy <olamy@apache.org>

See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).

### Proposed changelog entries

- Entry 1: Issue, human-readable text
- […]

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/8380"></a>

### CODE REVIEW

**Review by basil — CHANGES_REQUESTED** [review]
Incomplete—does not include the requisite WebSocket changes for Jetty 12.

File: pom.xml [hunk 1]
```diff
@@ -53,6 +53,7 @@ THE SOFTWARE.
     <module>bom</module>
     <module>websocket/spi</module>
     <module>websocket/jetty10</module>
+    <module>websocket/jetty12</module>
```
Comment (basil) [inline pom.xml:56]: For consistency with the existing `jetty10` module, the addition of this new module will require a corresponding change to `.idea/encodings.xml`:

```
    <file url="file://$PROJECT_DIR$/websocket/jetty12/src/filter/resources" charset="UTF-8" />
    <file url="file://$PROJECT_DIR$/websocket/jetty12/src/main/java" charset="UTF-8" />
    <file url="file://$PROJECT_DIR$/websocket/jetty12/src/main/resources" charset="UTF-8" />
```

File: websocket/jetty12/pom.xml [hunk 2]
```diff
@@ -0,0 +1,85 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+The MIT License
+
+Copyright (c) 2019, CloudBees, Inc.
```
Comment (basil) [inline websocket/jetty12/pom.xml:5]: 2024?

File: websocket/jetty12/src/main/java/jenkins/websocket/Jetty12Provider.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            JettyWebSocketServerContainer.getContainer(req.getServletContext()).setIdleTimeout(Duration.ofSeconds(IDLE_TIMEOUT_SECONDS));
+            initialized = true;
+        }
+    }
+
+    @Override
+    public Handler handle(HttpServletRequest req, HttpServletResponse rsp, Listener listener) throws Exception {
+        init(req);
+        req.setAttribute(ATTR_LISTENER, listener);
+        // TODO Jetty 10 has no obvious equivalent to WebSocketServerFactory.isUpgradeRequest; RFC6455Negotiation?
```
Comment (basil) [inline websocket/jetty12/src/main/java/jenkins/websocket/Jetty12Provider.java:77]: Since this class is focused on Jetty 12, it might be more accurate to write "Jetty 10+" here.

File: websocket/jetty12/src/main/java/jenkins/websocket/Jetty12Provider.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2022 CloudBees, Inc.
```
Comment (basil) [inline websocket/jetty12/src/main/java/jenkins/websocket/Jetty12Provider.java:4]: 2024?

File: websocket/jetty12/pom.xml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.jenkins-ci.main</groupId>
+    <artifactId>jenkins-parent</artifactId>
+    <version>${revision}${changelist}</version>
+    <relativePath>../..</relativePath>
+  </parent>
+
+  <artifactId>websocket-jetty12</artifactId>
```
Comment (basil) [inline websocket/jetty12/pom.xml:35]: Unlike previous versions, Jetty supports multiple servlet implementations, and in the future we want to add support for Jetty 12 with Jakarta imports. With this in mind, should we name this something like `websocket-jetty12-ee8`, so that it won't conflict with a future (potential) `websocket-jetty12-ee9` or `websocket-jetty12-ee10` module?

**Review by basil — CHANGES_REQUESTED** [review]
WebSocket test automation fails.

File: war/pom.xml [hunk 6]
```diff
@@ -632,9 +632,9 @@ THE SOFTWARE.
         </configuration>
       </plugin>
       <plugin>
-        <groupId>org.eclipse.jetty</groupId>
-        <artifactId>jetty-maven-plugin</artifactId>
-        <version>10.0.20</version>
+        <groupId>org.eclipse.jetty.ee8</groupId>
+        <artifactId>jetty-ee8-maven-plugin</artifactId>
+        <version>12.0.8</version>
```
Comment (basil) [inline war/pom.xml:637]: `jetty:run` Mojo fails with:

> [ERROR] Failed to execute goal `org.eclipse.jetty.ee8:jetty-ee8-maven-plugin:12.0.8:run` (`default-cli`) on project `jenkins-war`: Unable to parse configuration of mojo `org.eclipse.jetty.ee8:jetty-ee8-maven-plugin:12.0.8:run` for parameter `config`: Cannot create instance of class `org.eclipse.jetty.util.resource.Resource`: InstantiationException -> [Help 1]

Likely needs to adapt to https://github.com/jetty/jetty.project/pull/10077.

**Review by basil — CHANGES_REQUESTED** [review]

**Comment by basil** [comment]
Not only that, but the test harness needs significant changes for Jetty 12, along with these tests:

```
test/src/test/java/hudson/model/QueueTest.java
test/src/test/java/hudson/model/UpdateSiteTest.java
test/src/test/java/jenkins/install/SetupWizardTest.java
test/src/test/java/jenkins/util/SystemPropertiesTest.java
```

File: pom.xml (hunk 1 above)
Comment (olamy) [reply pom.xml:56]: TBH, having those ide files here is a really big pain. 
As I have winstone as a module in the same project,  some lines are added for winstone as well in this file.

File: websocket/jetty12/pom.xml (hunk 2 above)
Comment (olamy) [reply websocket/jetty12/pom.xml:5]: not sure really. as there are just copy of previous files with only few import changes.
so I cannot see real need to change that.

File: websocket/jetty12/pom.xml (hunk 5 above)
Comment (olamy) [reply websocket/jetty12/pom.xml:35]: sounds good idea.

File: .idea/encodings.xml [hunk 7]
```diff
@@ -32,8 +32,13 @@
     <file url="file://$PROJECT_DIR$/websocket/jetty10/src/filter/resources" charset="UTF-8" />
     <file url="file://$PROJECT_DIR$/websocket/jetty10/src/main/java" charset="UTF-8" />
     <file url="file://$PROJECT_DIR$/websocket/jetty10/src/main/resources" charset="UTF-8" />
+    <file url="file://$PROJECT_DIR$/websocket/jetty12/src/filter/resources" charset="UTF-8" />
+    <file url="file://$PROJECT_DIR$/websocket/jetty12/src/main/java" charset="UTF-8" />
+    <file url="file://$PROJECT_DIR$/websocket/jetty12/src/main/resources" charset="UTF-8" />
```
Comment (basil) [inline .idea/encodings.xml:37]: Inconsistent with the rest of this PR, which uses the module name `jetty12-ee8`:

```suggestion
    <file url="file://$PROJECT_DIR$/websocket/jetty12-ee8/src/filter/resources" charset="UTF-8" />
    <file url="file://$PROJECT_DIR$/websocket/jetty12-ee8/src/main/java" charset="UTF-8" />
    <file url="file://$PROJECT_DIR$/websocket/jetty12-ee8/src/main/resources" charset="UTF-8" />
```

File: .idea/encodings.xml [hunk 8]
```diff
@@ -32,8 +32,13 @@
...
     <file url="file://$PROJECT_DIR$/websocket/spi/src/filter/resources" charset="UTF-8" />
     <file url="file://$PROJECT_DIR$/websocket/spi/src/main/java" charset="UTF-8" />
     <file url="file://$PROJECT_DIR$/websocket/spi/src/main/resources" charset="UTF-8" />
+    <file url="file://$PROJECT_DIR$/../winstone/src/main/java" charset="UTF-8" />
+    <file url="file://$PROJECT_DIR$/../winstone/src/main/resources" charset="UTF-8" />
```
Comment (basil) [inline .idea/encodings.xml:42]: This doesn't seem essential to this PR, but rather an unrelated cleanup. I would suggest removing this hunk from this PR and submitting it as a separate change against the default branch, where it can be merged promptly rather than stagnating on this project branch that can't be merged for several weeks or months.

**Review by basil — CHANGES_REQUESTED** [review]
Still failing WebSocket `RealJenkinsRule` tests, presumably due to `websocket-jetty12-ee8` being missing from `war/pom.xml`.

File: websocket/jetty12/pom.xml (hunk 2 above)
Comment (basil) [reply websocket/jetty12/pom.xml]: It is common practice to preserve the created year and current year in the copyright notice; exempli gratia, "Copyright (c) 2019, 2024 CloudBees, Inc."

File: war/pom.xml (hunk 6 above)
Comment (basil) [reply war/pom.xml]: Fixed in commit d137f956b7

File: websocket/jetty12-ee8/pom.xml [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  <parent>
+    <groupId>org.jenkins-ci.main</groupId>
+    <artifactId>jenkins-parent</artifactId>
+    <version>${revision}${changelist}</version>
+    <relativePath>../..</relativePath>
+  </parent>
+
+  <artifactId>websocket-jetty12-ee8</artifactId>
+  <name>Jetty 12 implementation for WebSocket ee8</name>
+  <description>An implementation of the WebSocket handler that works with Jetty 12 based on EE8</description>
```
Comment (basil) [inline websocket/jetty12-ee8/pom.xml:37]: Inconsistent capitalization of "EE 8":

```suggestion
  <name>Jetty 12 (EE 8) implementation for WebSocket</name>
  <description>An implementation of the WebSocket handler that works with Jetty 12 (EE 8).</description>
```

File: websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12Provider.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+import org.eclipse.jetty.ee8.websocket.server.JettyServerUpgradeRequest;
+import org.eclipse.jetty.ee8.websocket.server.JettyServerUpgradeResponse;
+import org.eclipse.jetty.ee8.websocket.server.JettyWebSocketServerContainer;
+import org.kohsuke.MetaInfServices;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+@Restricted(NoExternalUse.class)
+@MetaInfServices(Provider.class)
+public class Jetty12Provider implements Provider {
```
Comment (basil) [inline websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12Provider.java:46]: Since this is now in a module named `jetty12-ee8`, and since we plan to eventually introduce another module for Jakarta imports (e.g., EE 9 or EE 10), I think this class should be named along the same lines, i.e., `Jetty12EE8Provider`.

File: websocket/jetty12/pom.xml (hunk 2 above)
Comment (olamy) [reply websocket/jetty12/pom.xml]: frankly it depends of projects.. Is it a common practice here? I'm not really sure 
I cannot see real modification of year when a file is modified here. 
looking at this commit https://github.com/jenkinsci/jenkins/commit/f3900d928ffc36dda47c5c6b0db416531e9543c1
I cannot see the copyright range modified for the pom

File: websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12E88Provider.java [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+import org.eclipse.jetty.ee8.websocket.server.JettyServerUpgradeRequest;
+import org.eclipse.jetty.ee8.websocket.server.JettyServerUpgradeResponse;
+import org.eclipse.jetty.ee8.websocket.server.JettyWebSocketServerContainer;
+import org.kohsuke.MetaInfServices;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+@Restricted(NoExternalUse.class)
+@MetaInfServices(Provider.class)
+public class Jetty12E88Provider implements Provider {
```
Comment (basil) [inline websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12E88Provider.java:46]: Typo?

```suggestion
public class Jetty12EE8Provider implements Provider {
```

File: websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12E88Provider.java [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+            JettyWebSocketServerContainer.getContainer(req.getServletContext()).setIdleTimeout(Duration.ofSeconds(IDLE_TIMEOUT_SECONDS));
+            initialized = true;
+        }
+    }
+
+    @Override
+    public Handler handle(HttpServletRequest req, HttpServletResponse rsp, Listener listener) throws Exception {
+        init(req);
+        req.setAttribute(ATTR_LISTENER, listener);
+        // TODO Jetty 12 has no obvious equivalent to WebSocketServerFactory.isUpgradeRequest; RFC6455Negotiation?
```
Comment (basil) [inline websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12E88Provider.java:77]: This comment is phrased as if it is specific to Jetty 12, but actually it is more general and applies to Jetty 10+ (and may very well apply to Jetty 13, etc in the future). To avoid having to keep updating the comment, why not change it from the particular to the generic:

```suggestion
        // TODO Jetty 10+ has no obvious equivalent to WebSocketServerFactory.isUpgradeRequest; RFC6455Negotiation?
```

That way the comment can remain as-is for years to come, until Jetty develops an obvious equivalent to the old method we used to rely on.

File: websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12E88Provider.java [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+
+@Restricted(NoExternalUse.class)
+@MetaInfServices(Provider.class)
+public class Jetty12E88Provider implements Provider {
+
+    /**
+     * Number of seconds a WebsocketConnection may stay idle until it expires.
+     * Zero to disable.
+     * This value must be higher than the <code>jenkins.websocket.pingInterval</code>.
+     * Per <a href=https://www.eclipse.org/jetty/documentation/jetty-12/programming-guide/index.html#pg-websocket-session-ping>Jetty 12 documentation</a>
```
Comment (basil) [inline websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12E88Provider.java:52]: Stale URL: this is now redirecting to a new URL, which is also linked from Jetty's official home page and GitHub page:

```suggestion
     * Per <a href=https://eclipse.dev/jetty/documentation/jetty-12/programming-guide/index.html#pg-websocket-session-ping>Jetty 12 documentation</a>
```

**Review by basil — COMMENTED** [review]
When upgrading Jetty to version 12 in Jenkins core, I think the core test suite should also be adapted to Jetty 12. Any progress on resolving the three known Jetty 12 test bugs in the core test suite? These are tracked in Jira:

- [JENKINS-73127](https://issues.jenkins.io/browse/JENKINS-73127)
- [JENKINS-73128](https://issues.jenkins.io/browse/JENKINS-73128)
- [JENKINS-73129](https://issues.jenkins.io/browse/JENKINS-73129)

File: websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12E88Provider.java (hunk 12 above)
Comment (olamy) [reply websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12E88Provider.java]: but it's part of Jetty12 related class..

File: websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12E88Provider.java (hunk 13 above)
Comment (olamy) [reply websocket/jetty12-ee8/src/main/java/jenkins/websocket/Jetty12E88Provider.java]: and soon redirect to another one https://jetty.org/..... 
a redirect is not stale as it's maintained.

**Comment by olamy** [comment]
As I have giving up with https://github.com/jenkinsci/winstone/pull/341 I close this PR as well.
Everybody is free to reuse the branch from the fork
